source/iofs/iofs.go

```diff
@@ -0,0 +1,164 @@
+// +build go1.16
+
+package iofs
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path"
+	"strconv"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+func init() {
+	source.Register("iofs", &IoFS{})
+}
+
+// IoFS is a source driver for io/fs#FS.
+type IoFS struct {
+	migrations *source.Migrations
+	fsys       fs.FS
+	path       string
+}
+
+// Open by url does not supported with IoFS.
+func (i *IoFS) Open(url string) (source.Driver, error) {
+	return nil, errors.New("iofs driver does not support open by url")
+}
+
+// WithInstance wraps io/fs#FS as source.Driver.
+func WithInstance(fsys fs.FS, path string) (source.Driver, error) {
+	var i IoFS
+	if err := i.Init(fsys, path); err != nil {
+		return nil, fmt.Errorf("failed to init driver with path %s: %w", path, err)
+	}
+	return &i, nil
+}
+
+// Init prepares not initialized IoFS instance to read migrations from a
+// fs.FS instance and a relative path.
+func (p *IoFS) Init(fsys fs.FS, path string) error {
```

Comment:
What are your thoughts about moving most of this implementation to a `PartialDriver`, similar to `httpfs`? We'd also need to rename `WithInstance()` to `New()` and change`IoFS` to `driver`. That way other drivers can embed the partial driver and implement their own `Open()` receiver function.